### PR TITLE
[SYCL][ESIMD] Deprecate block_load/store, add simd::copy_from/to.

### PR DESCRIPTION
**Pull request by kbobrovs**
This patch:
1) Fixes the following TODO in esimd_memory.hpp:
  // TODO @rolandschulz
  // Should follow existing std::simd naming for similar APIs - "copy_from" and
  // "copy_to" to avoid confusion.

2) Adds type checks for the sycl accessor arguments in the added APIs.

Signed-off-by: kbobrovs <Konstantin.S.Bobrovsky@intel.com>

### CODE REVIEW

**Comment by kbobrovs** [comment]
@rolandschulz, FYI

**Comment by kbobrovs** [comment]
End-to-end tests: https://github.com/intel/llvm-test-suite/pull/236

File: sycl/include/CL/sycl/INTEL/esimd/esimd.hpp [hunk 1]
```diff
@@ -480,6 +482,47 @@ template <typename Ty, int N> class simd {
...
+  /// global address space, otherwise behavior is undefined.
+  ESIMD_INLINE void copy_to(Ty *addr) SYCL_ESIMD_FUNCTION;
+
+  /// Copy all vector elements of this object into a contiguous block in memory.
+  /// Destination memory location is represented via a global accessor and
+  /// offset.
+  /// @param acc accessor to copy from.
+  /// @param offset offset to copy from.
+  template <typename AccessorT>
+  ESIMD_INLINE __ESIMD_ENABLE_IF_ACCESSOR(AccessorT, can_write, global_buffer,
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/INTEL/esimd/esimd.hpp:521]: NIT. Suggest aliasing the complex `enable_if` expression and use this alias here, like it's done with `EnableIfContiguous` in buffer.hpp:191.

File: sycl/test/esimd/simd_copy_to_copy_from.cpp [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+#ifdef __SYCL_DEVICE_ONLY__
+#define __SYCL_DEVICE_ATTR __attribute__((sycl_device))
+#else
+#define __SYCL_DEVICE_ATTR
+#endif // __SYCL_DEVICE_ONLY__
+
+// --- Postive tests.
+
+void kernel1(accessor<int, 1, access::mode::read_write,
```
Comment (DenisBakhvalov) [inline sycl/test/esimd/simd_copy_to_copy_from.cpp:24]: I think it's better to mark those functions with `SYCL_ESIMD_FUNCTION` even though it's not enforced now.

File: sycl/include/CL/sycl/INTEL/esimd/esimd.hpp (hunk 1 above)
Comment (kbobrovs) [reply sycl/include/CL/sycl/INTEL/esimd/esimd.hpp:521]: I think this one is more concise and readable, so I'd prefer this variant. Thanks for the NIT

File: sycl/include/CL/sycl/INTEL/esimd/detail/esimd_sycl_util.hpp [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  static_assert(Cap == accessor_mode_cap::can_read ||
+                    Cap == accessor_mode_cap::can_write,
+                "unsupported capability");
+
+  if constexpr (Mode == sycl::access::mode::atomic ||
+                Mode == sycl::access::mode::read_write ||
+                Mode == sycl::access::mode::discard_read_write)
+    return true; // atomic and *read_write accessors can read/write
+
+  return (Cap == accessor_mode_cap::can_read) ==
```
[the remark this answers is not in the document]
Comment (kbobrovs) [reply sycl/include/CL/sycl/INTEL/esimd/detail/esimd_sycl_util.hpp:62]: no, it is correct as written

File: sycl/include/CL/sycl/INTEL/esimd/detail/esimd_sycl_util.hpp [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+  static inline constexpr accessor_mode_cap_val_t can_write = true;
+};
+
+template <sycl::access::mode Mode, accessor_mode_cap_val_t Cap>
+constexpr bool accessor_mode_has_capability() {
+  static_assert(Cap == accessor_mode_cap::can_read ||
+                    Cap == accessor_mode_cap::can_write,
+                "unsupported capability");
+
+  if constexpr (Mode == sycl::access::mode::atomic ||
```
[the remark this answers is not in the document]
Comment (kbobrovs) [reply sycl/include/CL/sycl/INTEL/esimd/detail/esimd_sycl_util.hpp:57]: no, it is needed

File: sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp [hunk 5]
```diff
@@ -184,30 +185,20 @@ ESIMD_INLINE ESIMD_NODEBUG simd<T, n> block_load(const T *const addr) {
...
-  return __esimd_block_read<T, n>(acc, offset);
-#endif // __SYCL_DEVICE_ONLY__
+  simd<T, n> Res;
+  Res.copy_from(acc, offset);
+  return Res;
 }
 
 /// Flat-address block-store.
 /// \ingroup sycl_esimd
+// TODO the above note about cache hints applies to this API as well.
```
[the remark this answers is not in the document]
Comment (kbobrovs) [reply sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp:198]: I'd rather not duplicate the same text, which is not one-liner. Those APIs should later be processed in a bulk anyway.

File: sycl/include/CL/sycl/INTEL/esimd/esimd.hpp (hunk 1 above)
Comment (kbobrovs) [reply sycl/include/CL/sycl/INTEL/esimd/esimd.hpp:521]: looks like I looked at the wrong place. Point of usage seems good indeed. will do

File: sycl/test/esimd/simd_copy_to_copy_from.cpp (hunk 2 above)
Comment (kbobrovs) [reply sycl/test/esimd/simd_copy_to_copy_from.cpp:24]: right, thanks.

File: sycl/test/esimd/simd_copy_to_copy_from.cpp [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+#include <CL/sycl.hpp>
+#include <CL/sycl/INTEL/esimd.hpp>
+#include <limits>
+#include <utility>
+
+using namespace sycl::INTEL::gpu;
+using namespace cl::sycl;
+
+#ifdef __SYCL_DEVICE_ONLY__
+#define __SYCL_DEVICE_ATTR __attribute__((sycl_device))
```
[the remark this answers is not in the document]
Comment (kbobrovs) [reply sycl/test/esimd/simd_copy_to_copy_from.cpp:17]: good idea, thanks

File: sycl/include/CL/sycl/INTEL/esimd/esimd.hpp [hunk 7]
```diff
@@ -498,6 +541,86 @@ ESIMD_INLINE simd<U, n> convert(simd<T, n> val) {
   return __builtin_convertvector(val.data(), vector_type_t<U, n>);
 }
 
+// ----------- Outlined implementations of esimd class APIs.
+
+template <typename T, int N> void simd<T, N>::copy_from(const T *const addr) {
+  constexpr unsigned Sz = sizeof(T) * N;
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/INTEL/esimd/esimd.hpp:547]: Please, align style of variables naming: `addr` and `Sz` have different style.

File: sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp [hunk 8]
```diff
@@ -158,14 +158,15 @@ scatter(T *p, simd<T, n * ElemsPerAddr> vals, simd<uint32_t, n> offsets,
...
-//
 /// Flat-address block-load.
 /// \ingroup sycl_esimd
+// TODO normally, this function should just delegate to
+// simd::copy_from for the deprecation period, but separate implementations are
+// needed for now, as simd::copy_from does not support cache hints yet.
+// This API, even though deprecated, can't be removed until then.
 template <typename T, int n, CacheHint L1H = CacheHint::None,
           CacheHint L3H = CacheHint::None>
+__SYCL_DEPRECATED("Replaced by simd::copy_from")
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp:169]: NIT. Consider rephrasing to "XYZ is deprecated. Use XZZ instead." to align with other deprecation warnings.

File: sycl/include/CL/sycl/INTEL/esimd/esimd.hpp (hunk 7 above)
Comment (kbobrovs) [reply sycl/include/CL/sycl/INTEL/esimd/esimd.hpp:547]: ok. Looks like we need  some update to the coding style, as SYCL public API spec uses lower case style, but we use LLVM style internally, sometimes those contradict.

File: sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp (hunk 8 above)
Comment (kbobrovs) [reply sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp:169]: ok

**Comment by kbobrovs** [comment]
had to rebase and force-push to resolve conflicts.

@romanovvlad, @DenisBakhvalov - sorry for delay. Please review.

File: sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp [hunk 9]
```diff
@@ -158,14 +158,15 @@ scatter(T *p, simd<T, n * ElemsPerAddr> vals, simd<uint32_t, n> offsets,
...
-//
 /// Flat-address block-load.
 /// \ingroup sycl_esimd
+// TODO normally, this function should just delegate to
+// simd::copy_from for the deprecation period, but separate implementations are
+// needed for now, as simd::copy_from does not support cache hints yet.
+// This API, even though deprecated, can't be removed until then.
 template <typename T, int n, CacheHint L1H = CacheHint::None,
           CacheHint L3H = CacheHint::None>
+__SYCL_DEPRECATED("block_load is deprecated, use simd::copy_from.")
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp:169]: @romanovvlad, I need to take this back. XXX is deprecated, use XYZ instead leads to unneeded duplication in the warning message:

```
D:\BuildBot\Product_worker_intel\sycl-win-x64-pr\llvm.src\sycl\test\esimd\odr.cpp:38:30: warning: 'block_load<int, 8, sycl::INTEL::gpu::CacheHint::None, sycl::INTEL::gpu::CacheHint::None>' is deprecated: block_load is deprecated, use simd::copy_from. [-Wdeprecated-declarations]
1 warning and 16 errors generated.
```

**Review by DenisBakhvalov — DISMISSED** [review]

**Comment by kbobrovs** [comment]
@DenisBakhvalov - I saw you reviewed 1 hour ago, somehow I did not see any comments. Were there any? If none - please approve.

@romanovvlad - please approve if there a no other comments.

**Review by DenisBakhvalov — APPROVED** [review]

**Review by romanovvlad — APPROVED** [review]

**Pull request merged**